pkg/volumes/openstack/util.go

```diff
@@ -0,0 +1,46 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openstack
+
+import (
+	"fmt"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+)
+
+const (
+	openstackExternalIPType = "OS-EXT-IPS:type"
+	openstackAddressFixed   = "fixed"
+	openstackAddress        = "addr"
+)
+
+func GetServerFixedIP(server *servers.Server) (poolAddress string, err error) {
+	for _, address := range server.Addresses {
+		if addresses, ok := address.([]interface{}); ok {
+			for _, addr := range addresses {
+				addrMap := addr.(map[string]interface{})
+				if addrType, ok := addrMap[openstackExternalIPType]; ok && addrType == openstackAddressFixed {
+					if fixedIP, ok := addrMap[openstackAddress]; ok {
+						if fixedIPStr, ok := fixedIP.(string); ok {
+							return fixedIPStr, nil
```

Comment:
This, like the google impl, assumes that the first interface with a fixed IP is the one we want.  This really doesn't make sense to me, but there is some information I cant assume here.  

It is likely that our router will be named `cluster-name` and that this would be the interface key in the server.Addresses map, however I only can assume that is true in the scope of a Kops deployment.

Reply:
imo this is good enough for now. Kops (where this is going to be used) does not support attaching more than one interface per instance. Maybe we could add comment about this beginning of this function? Not perfect solution if some day we want to support more than one interface